Inferences/SubsumptionDemodulationHelper.cpp

```diff
@@ -0,0 +1,278 @@
+    //
+    // NOTE: if there is more than one base literal without alts,
+    // subsumption resolution might still be possible
+    // (if we can unify them and resolve them all away)
+    for (unsigned i = 0; i < mcl->length(); ++i) {
+      if (LiteralList::isEmpty(cm.alts()[i])) {
+        if (!MatchingUtils::match((*mcl)[i], resLit, /* complementary = */ true)) {
+          return false;
+        }
+      }
+    }
+
+  } else {
+    // No base literal without alts?
+    // Subsumption resolution is possible if at least one base lit can be resolved with resLit
+
+    bool anyResolvable = false;
+    for (unsigned i = 0; i < mcl->length(); ++i) {
+      if (MatchingUtils::match((*mcl)[i], resLit, /* complementary = */ true)) {
+        anyResolvable = true;
+        break;
+      }
+    }
+    if (!anyResolvable) {
+      return false;
+    }
+  }
+
+  // NOTE: we use MLMatcher here because we want *subset* inclusion (as opposed to submultiset)
+  return MLMatcher::canBeMatched(mcl, cl, cm.alts(), resLit);
+}
+/**
+ * Build clause that results from subsumption resolution with main premise 'cl' and side premise 'mcl'.
+ * The literal 'resLit' is the resolved literal from 'cl'.
+ */
+Clause* SDHelper::generateSubsumptionResolutionClause(Clause* cl, Literal* resLit, Clause* mcl)
+{
+  CALL("generateSubsumptionResolutionClause");
+
+  unsigned newLen = cl->length() - 1;
+  Clause* newCl = new(newLen) Clause(newLen,
+      SimplifyingInference2(InferenceRule::SUBSUMPTION_RESOLUTION, cl, mcl));
+
+  unsigned j = 0;
+  for (unsigned i = 0; i < cl->length(); ++i) {
+    Literal* curLit = (*cl)[i];
+
+    if (curLit != resLit) {
+      (*newCl)[j] = curLit;
+      j += 1;
+    }
+  }
+  // We should have skipped exactly one literal, namely resLit.
+  // (it should never appear twice because we apply duplicate literal removal before subsumption resolution)
+  ASS_EQ(j, newLen);
+
+  return newCl;
+}
+
+
+#if VDEBUG
+/// Returns true iff clause with literal lits1 is smaller than clause with literals lits2
+/// in the multiset extension of the given ordering.
+///
+/// This implementation is justified by Lemma 2.5.6 on page 24 of [BN98].
+/// [BN98] Franz Baader and Tobias Nipkow. Term Rewriting and All That. Cambridge University Press, 1998.
+SDHelper::ClauseComparisonResult SDHelper::clauseCompare(Literal* const lits1[], unsigned n1, Literal* const lits2[], unsigned n2, Ordering const& ordering)
+{
+  // "-lcm reverse" messes up the ordering
+  ASS(env.options->literalComparisonMode() != Options::LiteralComparisonMode::REVERSE);
+
+  // Copy given literals so we can sort them
+  vvector<Literal*> c1(lits1, lits1+n1);
+  vvector<Literal*> c2(lits2, lits2+n2);
+
+  // These will contain literals from c1/c2 with equal occurrences removed
+  vvector<Literal*> v1;
+  vvector<Literal*> v2;
+
+  // The equality tests below only make sense for shared literals
+  std::for_each(c1.begin(), c1.end(), [](Literal* lit) { ASS(lit->shared()); });
```

Comment:
Will this for_each run even in release mode? Perhaps it should be inside a VDEBUG